src/coreclr/vm/methodtable.cpp

```diff
@@ -3483,41 +3484,75 @@ void MethodTable::AllocateRegularStaticBoxes()
     PTR_BYTE pStaticBase = GetGCStaticsBasePointer();
 
     GCPROTECT_BEGININTERIOR(pStaticBase);
-
     {
         FieldDesc *pField = HasGenericsStaticsInfo() ?
             GetGenericsStaticFieldDescs() : (GetApproxFieldDescListRaw() + GetNumIntroducedInstanceFields());
         FieldDesc *pFieldEnd = pField + GetNumStaticFields();
 
         while (pField < pFieldEnd)
         {
-            _ASSERTE(pField->IsStatic());
-
             if (!pField->IsSpecialStatic() && pField->IsByValue())
             {
-                TypeHandle  th = pField->GetFieldTypeHandleThrowing();
-                MethodTable* pFieldMT = th.GetMethodTable();
-
-                LOG((LF_CLASSLOADER, LL_INFO10000, "\tInstantiating static of type %s\n", pFieldMT->GetDebugClassName()));
-                OBJECTREF obj = AllocateStaticBox(pFieldMT, HasFixedAddressVTStatics());
-
-                SetObjectReference( (OBJECTREF*)(pStaticBase + pField->GetOffset()), obj);
+                AllocateRegularStaticBox(pField, pStaticBase);
             }
-
             pField++;
         }
     }
     GCPROTECT_END();
 }
 
+void MethodTable::AllocateRegularStaticBox(FieldDesc* pField, BYTE* pStaticBase)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+        CONTRACTL_END;
+    }
+    _ASSERT(pField->IsStatic() && !pField->IsSpecialStatic() && pField->IsByValue());
+
+    Object** boxedStaticHandle = (Object**)(pStaticBase + pField->GetOffset());
+
+    if (VolatileLoad(boxedStaticHandle) != nullptr)
+    {
+        // Boxed static is already initialized
+        return;
+    }
+
+    // Taking a lock since we might come here from multiple threads/places
+    CrstHolder crst(GetAppDomain()->GetStaticBoxInitLock());
+
+    // double-checked locking
+    if (VolatileLoad(boxedStaticHandle) != nullptr)
+    {
+        // Boxed static is already initialized
+        return;
+    }
+
+    TypeHandle  th = pField->GetFieldTypeHandleThrowing();
```

Comment:
`GetFieldTypeHandleThrowing` can run a type loader. We should not run type loader with the lock taken.

(The problem was likely hidden by using `Unordered` as the lock order.)